src/strings.rs

```diff
@@ -0,0 +1,230 @@
+//!
+//! [ascii]: https://numpy.org/doc/stable/reference/c-api/dtype.html#c.NPY_STRING
+//! [ucs4]: https://numpy.org/doc/stable/reference/c-api/dtype.html#c.NPY_UNICODE
+
+use std::cell::RefCell;
+use std::collections::hash_map::Entry;
+use std::convert::TryInto;
+use std::fmt;
+use std::mem::size_of;
+use std::os::raw::c_char;
+use std::str;
+
+use pyo3::{
+    ffi::{Py_UCS1, Py_UCS4},
+    sync::GILProtected,
+    Py, Python,
+};
+use rustc_hash::FxHashMap;
+
+use crate::dtype::{Element, PyArrayDescr};
+use crate::npyffi::NPY_TYPES;
+
+/// A newtype wrapper around [`[u8; N]`][Py_UCS1] to handle [`byte` scalars][numpy-bytes] while satisfying coherence.
+///
+/// Note that when creating arrays of ASCII strings without an explicit `dtype`,
+/// NumPy will automatically determine the smallest possible array length at runtime.
+///
+/// For example,
+///
+/// ```python
+/// array = numpy.array([b"foo", b"bar", b"foobar"])
+/// ```
+///
+/// yields `S6` for `array.dtype`.
+///
+/// On the Rust side however, the length `N` of `PyFixedString<N>` must always be given
+/// explicitly and as a compile-time constant. For this work reliably, the Python code
+/// should set the `dtype` explicitly, e.g.
+///
+/// ```python
+/// numpy.array([b"foo", b"bar", b"foobar"], dtype='S12')
+/// ```
+///
+/// always matching `PyArray1<PyFixedString<12>>`.
+///
+/// # Example
+///
+/// ```rust
+/// # use pyo3::Python;
+/// use numpy::{PyArray1, PyFixedString};
+///
+/// # Python::with_gil(|py| {
+/// let array = PyArray1::<PyFixedString<3>>::from_vec(py, vec![[b'f', b'o', b'o'].into()]);
+///
+/// assert!(array.dtype().to_string().contains("S3"));
+/// # });
+/// ```
+///
+/// [numpy-bytes]: https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.bytes_
+#[repr(transparent)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct PyFixedString<const N: usize>(pub [Py_UCS1; N]);
+
+impl<const N: usize> fmt::Display for PyFixedString<N> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.write_str(str::from_utf8(&self.0).unwrap().trim_end_matches('\0'))
+    }
+}
+
+impl<const N: usize> From<[Py_UCS1; N]> for PyFixedString<N> {
+    fn from(val: [Py_UCS1; N]) -> Self {
+        Self(val)
+    }
+}
+
+unsafe impl<const N: usize> Element for PyFixedString<N> {
+    const IS_COPY: bool = true;
+
+    fn get_dtype(py: Python) -> &PyArrayDescr {
+        static DTYPES: TypeDescriptors = TypeDescriptors::new();
+
+        unsafe { DTYPES.from_size(py, NPY_TYPES::NPY_STRING, b'|' as _, size_of::<Self>()) }
+    }
+}
+
+/// A newtype wrapper around [`[PyUCS4; N]`][Py_UCS4] to handle [`str_` scalars][numpy-str] while satisfying coherence.
+///
+/// Note that when creating arrays of Unicode strings without an explicit `dtype`,
+/// NumPy will automatically determine the smallest possible array length at runtime.
+///
+/// For example,
+///
+/// ```python
+/// numpy.array(["foo", "bar", "foobar"])
```

Comment:
This line looks unchanged after you put 🐍 and 🦀 in the code snippet below.

Reply:
Indeed. Fixed now.